### PR TITLE
Add configuration to compile Magick.Native on arm64

### PR DESCRIPTION
**Pull request by alastairporter**
Hi!
Following the conversation on https://github.com/dlemstra/Magick.NET/issues/1000, I've managed to get Magick.Native compiling natively on an apple silicon machine. I ran this on a brand new machine with MacOS 12 installed, not a github runner (which I understand comes with a few tools already installed). This required a few small changes and fixes. I added comments to the git commits with more details, but also include some notes here:

On apple silicon, homebrew installs into /opt, not /usr/local, and /usr/local is protected by SIP, meaning that you cannot chown the directory to the user that you're running as. I ended up adding some directories and chowning them manually:

    sudo mkdir -p /usr/local/{bin,lib,share/man/man3,include,etc,var}
    sudo chown -R myuser /usr/local/*

I'm not sure where the best place would be to put this in the build scripts, especially because of the requirement of `sudo`.

A base macOS install comes with only python3/pip3, which means I updated some `pip install` commands in `build.fontconfig.sh`. Additionally, system pip3 by default installs packages to a user's local Library, rather than a system-wide directory, which means that I had to manually modify the PATH to find `meson` after installing it:

    export PATH=/Users/m1/Library/Python/3.8/bin:$PATH

On Apple Silicon, fontconfig crashes when running `fc-cache`. This appears to be fixed by https://gitlab.freedesktop.org/fontconfig/fontconfig/-/merge_requests/88, I applied these changes manually on the local checkout of fontconfig and it successfully compiled and installed.

The contents of the `build/macos-arm64/` directory is almost identical to the x64 version, except for a change in the `FLAGS` env variable. Maybe there's a better way of ensuring that these files are kept in sync, perhaps with symlinks, or a single common file? Happy to discuss possibilities here.

## Next steps

I've also looked into cross-compiling arm64 libraries on an x86 runner (and therefore being able to run this build process on a github runner). As far as I can tell, this should be as simple as adding `-target arm64-apple-macos11 ` to the CFLAGS/CXXFLAGS as needed. I wasn't sure how we could configure this with the existing setup. Does this mean we should add a new `build/macos-x64-cross-arm64` directory with _another_ copy of the settings and build scripts? If we can come up with a preferred directory structure, I can add this config too, which means that we can build a native version of Magick.Native for arm64 and then make it available for Magick.Net too.

Happy to answer any questions, thanks!

### CODE REVIEW

**Comment by dlemstra** [comment]
Thanks for your pull request and your explanation. I think it would be best to do a cross-platform compilation. This is also done for the `linux-arm64` configuration. The testing for this project is done inside Magick.NET and that should be added once arm64 works for macos through GitHub actions. And yes there are a lot of duplicate lines between those folders but I don't mind the duplication in this case. It might be a good idea to check what is done for the `linux-arm64` build when you add this cross-compilation. Feel free to reach out to me if you have any further questions.

**Comment by alastairporter** [comment]
OK, Great. I'll configure the `macos-arm64` directory to do cross-compilation from an x64 machine instead of native compilation, and will add this to the runner configuration for you to approve and test.

**Comment by omidontop** [comment]
Any progress? Can I help in any way?

**Comment by alastairporter** [comment]
@dlemstra I've updated this to build the apple silicon version on an x86-64 mac. I tested the build scripts on a new empty VM and they compiled successfully and the collected artifacts that are generated are for arm:

```
 % file artifacts/ReleaseQ16/arm64/Magick.Native-Q16-arm64.dll.dylib
artifacts/ReleaseQ16/arm64/Magick.Native-Q16-arm64.dll.dylib: Mach-O 64-bit dynamically linked shared library arm64
```

I rebased this from `main` to include the recent changes that you made to the x86 install scripts.

This requires one additional fix to a dependency: https://github.com/ImageMagick/glib/pull/1, I see that the clone script loads a commit close to the age of the ImageMagick sub repository, so you may need to update this to a newer version to make the patch apply. Alternatively, we can keep a .patch file in this repo and apply it before building, let me know what you prefer.

I added a section to the workflow script to build these, but I'm not sure how to run the "verify" jobs that you have set up, as I guess we need a worker running on apple silicon. Perhaps you have a recommendation about how we could do this.

@omidontop Thanks for the offer to help on building these, but I think everything is working for now. Let's see how the build/packaging process goes and we can see if there's any more work to be done.

Thanks!

**Comment by alastairporter** [comment]
@robrab2000 I'm actually not sure why the [mac build script](https://github.com/dlemstra/Magick.Native/blob/611eae6bd469168963c9f2d300326eab7b8c5864/build/macos-x64/install.dependencies.sh#L61) tries to install ninja 1.10.0.post2, even though the [linux ones install a later version](https://github.com/dlemstra/Magick.Native/blob/611eae6bd469168963c9f2d300326eab7b8c5864/build/linux-x64/install.dependencies.sh#L11). Given that you're using python 3.10, my guess is that you were unable to download the precompiled wheel and so tried to compile ninja from scratch. I recommend that you try and upgrade to the latest version of ninja (1.10.2.4) and see if it can install the wheel. We should double-check the macos and python versions in the github runners to see if we can also get away with upgrading a newer version.

Just a small note to keep in mind that the install scripts are pretty aggressive and assume that they're the only thing running on the machine, they uninstall a bunch of packages from homebrew which you might not expect - make sure you look at the script to be sure about what it's doing.

**Comment by robrab2000** [comment]
Hi! Oh right! Thanks, sorry I'm a Unity dev so working with non-Unity libraries is all pretty new to me.

Thanks for the heads up on the uninstalling stuff, I've gone through and reinstalled all my brew packages that got uninstalled.

I loaded the `Magic.NET.Core` and `Magic.NET-Q16-HDRI-AnyCPU` packages from NuGet and when I tried to access them, I got this error: `"DllNotFoundException: Magick.Native-Q16-HDRI-x64.dll"`

From googling around it seemed like I needed to compile the native binary for my M1 and that's how I wound up here :)

Running the build script in a VM sounds like a much better plan, I've used VirtualBox before but I'm not sure this can virtualise with MacOS. Sorry if this is a noob question but what software should I be using for this. I thought of VMWare Fusion but it looks like this doesn't support M1 Macs. Parallels is an option too but I think this only virtualises Windows?

**Comment by omidontop** [comment]
> Hi! Oh right! Thanks, sorry I'm a Unity dev so working with non-Unity libraries is all pretty new to me.
> 
> Thanks for the heads up on the uninstalling stuff, I've gone through and reinstalled all my brew packages that got uninstalled.
> 
> I loaded the `Magic.NET.Core` and `Magic.NET-Q16-HDRI-AnyCPU` packages from NuGet and when I tried to access them, I got this error: `"DllNotFoundException: Magick.Native-Q16-HDRI-x64.dll"`
> 
> From googling around it seemed like I needed to compile the native binary for my M1 and that's how I wound up here :)
> 
> Running the build script in a VM sounds like a much better plan, I've used VirtualBox before but I'm not sure this can virtualise with MacOS. Sorry if this is a noob question but what software should I be using for this. I thought of VMWare Fusion but it looks like this doesn't support M1 Macs. Parallels is an option too but I think this only virtualises Windows?

Parallels can virtualize any OS including Windows and Linux. The problem with Parallels and VMWare Fusion is that none of them can virtualize OSes compiled for other processor architectures. So your VM will also be an arm64 environment. I've read that https://getutm.app/ can emulate an x64 or another architecture on M1 but have never used it myself.

**Comment by robrab2000** [comment]
Thanks, I've set up a macOS vm on parallels and managed to run the dependencies script!

Next I've run the `build.libraries.sh` script but encountered this error:

`"./build.libraries.sh: line 9: /build.zlib.sh: No such file or directory"`

Do I need to put zlib in the folder or something?

**Comment by alastairporter** [comment]
@dlemstra I'm not sure what's going on here, I see the cairo build error message for the arm only build:

```
2022-10-11T14:52:52.7643960Z In file included from cairo-xlib-fallback-compositor.c:47:
2022-10-11T14:52:52.7644360Z In file included from ./cairo-xlib-private.h:40:
2022-10-11T14:52:52.7645010Z ./cairo-xlib-xrender-private.h:132:3: error: typedef redefinition with different types ('struct (anonymous struct at ./cairo-xlib-xrender-private.h:129:16)' vs 'struct _XConicalGradient')
2022-10-11T14:52:52.7645400Z } XConicalGradient;
2022-10-11T14:52:52.7645580Z   ^
2022-10-11T14:52:52.7646080Z /usr/local/include/X11/extensions/Xrender.h:199:3: note: previous definition is here
2022-10-11T14:52:52.7646400Z } XConicalGradient;
```

which kind of implies to me that this file was already in place before we tried to compile (another package provided by default on the github runners?). I definitely didn't see this error when I was building on a fresh macos installation. It's weird that it occurs on the arm64 build but not the x64 one though...

**Comment by alastairporter** [comment]
Inspecting the output of `./configure` for cairo, I see this:

For x64:
> 2022-10-11T14:34:37.2687940Z checking for X... libraries , headers 

For arm64:
> 2022-10-11T14:51:52.7897730Z checking for X... no

I guess what's going on here is that the .dylib file for X (whatever this check is) is built for x86 and not arm, and so it doesn't (or does add) some different compiler flags, causing an error during compilation (https://github.com/ImageMagick/cairo/blob/main/configure.ac).

I note that the output of configure is the same for both versions:
```
2022-10-11T14:49:28.6669730Z cairo (version 1.16.0 [release]) will be compiled with:
2022-10-11T14:49:28.6669920Z 
2022-10-11T14:49:28.6670520Z The following surface backends:
2022-10-11T14:49:28.6670810Z   Image:         yes (always builtin)
2022-10-11T14:49:28.6671040Z   Recording:     yes (always builtin)
2022-10-11T14:49:28.6671270Z   Observer:      yes (always builtin)
2022-10-11T14:49:28.6671490Z   Mime:          yes (always builtin)
2022-10-11T14:49:28.6671910Z   Tee:           no (disabled, use --enable-tee to enable)
2022-10-11T14:49:28.6672900Z   XML:           no (disabled, use --enable-xml to enable)
2022-10-11T14:49:28.6673140Z   Xlib:          yes
2022-10-11T14:49:28.6673430Z   Xlib Xrender:  yes
2022-10-11T14:49:28.6673770Z   Qt:            no (disabled, use --enable-qt to enable)
2022-10-11T14:49:28.6674150Z   Quartz:        no (disabled, use --enable-quartz to enable)
2022-10-11T14:49:28.6674570Z   Quartz-image:  no (disabled, use --enable-quartz-image to enable)
2022-10-11T14:49:28.6674840Z   XCB:           yes
2022-10-11T14:49:28.6675060Z   Win32:         no (requires a Win32 platform)
2022-10-11T14:49:28.6675430Z   OS2:           no (disabled, use --enable-os2 to enable)
2022-10-11T14:49:28.6675670Z   CairoScript:   yes
2022-10-11T14:49:28.6675860Z   PostScript:    yes
2022-10-11T14:49:28.6676060Z   PDF:           yes
2022-10-11T14:49:28.6676240Z   SVG:           yes
2022-10-11T14:49:28.6676570Z   OpenGL:        no (disabled, use --enable-gl to enable)
2022-10-11T14:49:28.6676970Z   OpenGL ES 2.0: no (disabled, use --enable-glesv2 to enable)
2022-10-11T14:49:28.6677350Z   OpenGL ES 3.0: no (disabled, use --enable-glesv3 to enable)
2022-10-11T14:49:28.6677730Z   BeOS:          no (disabled, use --enable-beos to enable)
2022-10-11T14:49:28.6678100Z   DirectFB:      no (disabled, use --enable-directfb to enable)
2022-10-11T14:49:28.6678480Z   OpenVG:        no (disabled, use --enable-vg to enable)
2022-10-11T14:49:28.6678840Z   DRM:           no (disabled, use --enable-drm to enable)
2022-10-11T14:49:28.6679210Z   Cogl:          no (disabled, use --enable-cogl to enable)
2022-10-11T14:49:28.6679360Z 
2022-10-11T14:49:28.6679460Z The following font backends:
2022-10-11T14:49:28.6679680Z   User:          yes (always builtin)
2022-10-11T14:49:28.6679880Z   FreeType:      yes
2022-10-11T14:49:28.6680070Z   Fontconfig:    yes
2022-10-11T14:49:28.6680290Z   Win32:         no (requires a Win32 platform)
2022-10-11T14:49:28.6680650Z   Quartz:        no (disabled, use --enable-quartz to enable)
2022-10-11T14:49:28.6680940Z 
2022-10-11T14:49:28.6681030Z The following functions:
2022-10-11T14:49:28.6681240Z   PNG functions:   yes
2022-10-11T14:49:28.6681490Z   GLX functions:   no (not required by any backend)
2022-10-11T14:49:28.6681740Z   WGL functions:   no (not required by any backend)
2022-10-11T14:49:28.6682000Z   EGL functions:   no (not required by any backend)
2022-10-11T14:49:28.6682420Z   X11-xcb functions: no (disabled, use --enable-xlib-xcb to enable)
2022-10-11T14:49:28.6683170Z   XCB-shm functions: yes
2022-10-11T14:49:28.6683310Z 
2022-10-11T14:49:28.6683410Z The following features and utilities:
2022-10-11T14:49:28.6683730Z   cairo-trace:                yes
2022-10-11T14:49:28.6684020Z   cairo-script-interpreter:   yes
2022-10-11T14:49:28.6684180Z 
2022-10-11T14:49:28.6684280Z And the following internal features:
2022-10-11T14:49:28.6684500Z   pthread:       yes
2022-10-11T14:49:28.6684750Z   gtk-doc:       no
2022-10-11T14:49:28.6684960Z   gcov support:  no
2022-10-11T14:49:28.6685350Z   symbol-lookup: no (requires bfd)
2022-10-11T14:49:28.6685740Z   test surfaces: no (disabled, use --enable-test-surfaces to enable)
2022-10-11T14:49:28.6686020Z   ps testing:    no (requires libspectre)
2022-10-11T14:49:28.6686390Z   pdf testing:   no (requires poppler-glib >= 0.17.4)
2022-10-11T14:49:28.6686750Z   svg testing:   no (requires librsvg-2.0 >= 2.35.0)
```
despite the fact that the arm version failed to find X, it still seems happy to build the xlib backends. I don't know if these backends are actually necessary for cairo/imagemagick when running on a mac.

I expect that the fix to this would be one of:
 * uninstall libxrender from brew before trying to compile so that the backend isn't compiled
 * add `--disable-xlib` to the configure commandline to disable it
 * compile xlib/xlibrender from source during the build step so that we can build an arm64 version for the cross compile

**Comment by dlemstra** [comment]
I have added `--disable-xlib` to the build options not sure if that helps but probably a good idea to add it.

**Comment by alastairporter** [comment]
Thanks, I just merged `main` into here to see if that works.
In [your workflow on main](https://github.com/dlemstra/Magick.Native/actions/runs/3229896248/jobs/5287712718) I now see xlib disabled, even though it found the libraries:
```
2022-10-11T20:19:43.2161030Z checking for X... libraries , headers 
```

```
2022-10-11T20:33:45.9172700Z   Xlib:          no (disabled, use --enable-xlib to enable)
2022-10-11T20:33:45.9173060Z   Xlib Xrender:  no (requires --enable-xlib)
```

**Comment by alastairporter** [comment]
Thanks for kicking off the run.
It looks like this failure is in gfk-pixbuf, as part of the [rsvg](https://github.com/dlemstra/Magick.Native/blob/main/build/libraries/build.rsvg.sh) build script.
The specific error is that it can't compile a test program with libjpeg:

arm64
```
2022-10-12T06:37:22.6010530Z checking whether to build gmodulized gdk-pixbuf... no
2022-10-12T06:37:22.6497990Z checking for libtiff-4... yes
2022-10-12T06:37:22.7870790Z checking for jpeg_destroy_decompress in -ljpeg... no
    ... [failure]
```
x64:
```
2022-10-12T06:46:26.3868910Z checking whether to build gmodulized gdk-pixbuf... no
2022-10-12T06:46:26.4019010Z checking for libpng16... yes
2022-10-12T06:46:26.4149910Z checking pixbuf loaders to build... ani,icns,tga,png,pnm,xbm,xpm,qtif,bmp,gif,ico,jpeg,tiff
```

What's interesting to me is the order of the tests in the configure script. I see that the x64 one checks for libpng first whereas the cross compile one checks for libtiff and libjpeg.
I see in the configure script there is a guard around checking these two libraries, if the os is windows or `--without-gdiplus` is set: https://github.com/ImageMagick/rsvg/blob/f932b994bf0215518459e1e6b19e4f88e07736dd/gdk-pixbuf/configure.ac#L457, which makes me suspect that something funny is happening in the configure script.

I've tested again on my empty VM and this library compiles cleanly, so I suspect there's another conflict with existing homebrew packages on the github runners. I think that I might need to run some tests directly on the runner to see what's happening during compilation, I'll see if I can do this.

Again, an alternative here would be to disable these loaders in gdk-pixbuf, but I'm unsure if they're needed for anything in rsvg/imagemagick.

**Comment by alastairporter** [comment]
OK, I did some debugging into this to see what's going on.

The failure 
```
2022-10-12T06:37:22.7870790Z checking for jpeg_destroy_decompress in -ljpeg... no
```
is because we had a version of `/usr/local/lib/libjpeg.dylib` installed from homebrew:
```
lrwxr-xr-x    1 runner  admin    44 Oct  3 02:34 libjpeg.dylib -> ../Cellar/jpeg-turbo/2.1.4/lib/libjpeg.dylib
```
and during `./configure` for gdk-pixbuf it tries to compile against `-ljpeg`, causing the following error (from config.log):
```
2022-10-12T13:16:50.8819970Z configure:16809: gcc -o conftest -O3 -fPIC -arch arm64 -Wall -I/usr/local/include -DG_DISABLE_SINGLE_INCLUDES -L/usr/local/lib conftest.c -ljpeg   >&5
2022-10-12T13:16:50.8820710Z ld: warning: ignoring file /usr/local/lib/libjpeg.dylib, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
```

I added a `brew uninstall jpeg-turbo` to `install.dependencies.sh` and now gdk-pixbuf successfully compiles.

However, my other intuition was also correct and I see the following logging in config.log:

x86_64:
```
GDK_PIXBUF_EXTRA_LIBS='  -L/usr/local/lib -lpng16    /usr/local/lib/libtiff.la'
```

arm64:
```
GDK_PIXBUF_EXTRA_LIBS=' -ljpeg -L/usr/local/lib -lpng16  -lm  /usr/local/lib/libtiff.la'
```

So the configure script appears to be skipping the test for libjpeg in the x86_64 build: https://github.com/ImageMagick/rsvg/blob/f932b994bf0215518459e1e6b19e4f88e07736dd/gdk-pixbuf/configure.ac#L472-L473

I'm not sure why this is happening, but it appears to be something relevant to the existing mac x86-64 build, I had a look through build logs from a month ago and it was also skipping this library. The linux builds seem to correctly perform this test (both the x64 and the arm cross compile build).

**Comment by alastairporter** [comment]
@dlemstra this looks great, thanks for running the builds. Is there anything else we need to do to to get this merged?

**Comment by dlemstra** [comment]
I haven't done a review of the code yet. Will probably do that this weekend. And before we can use this in Magick.NET we also need to make a couple code changes there but I will do that after we get this merged here.

**Comment by dlemstra** [comment]
Not sure if it is possible but maybe we could do some kind of verification of the build? Do some check that the created binary really is arm64? For the Linux build I use qemu to verify this but maybe we could execute something to verify that it are really arm64 binaries. And I would prefer to do that in a separate job in the build. We can also do that in a follow up PR.

**Comment by zongjingyao** [comment]
@dlemstra You can use `file` command to check the binary's architecture.
image

**Comment by alastairporter** [comment]
I agree, `file [thefile]` or `lipo -info [thefile]` is going to directly tell you the architecture of the file. 

I see that in the `verify.Native.sh` script for linux you also check if it's linked against openmp in the case that it's an openmp variant. On a mac you can use `otool -L` for something similar:

```
alastair@macos-12 arm64 % file Magick.Native-Q16-arm64.dll.dylib
Magick.Native-Q16-arm64.dll.dylib: Mach-O 64-bit dynamically linked shared library arm64
alastair@macos-12 arm64 % lipo -info Magick.Native-Q16-arm64.dll.dylib
Non-fat file: Magick.Native-Q16-arm64.dll.dylib is architecture: arm64
alastair@macos-12 arm64 % otool -L Magick.Native-Q16-arm64.dll.dylib
Magick.Native-Q16-arm64.dll.dylib:
	@rpath/libMagick.Native-Q16-arm64.dll.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1300.23.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)
```

I see that you're using `ld` to try and "link" the shared library to find out if there are any unknown symbols. I'm not sure how you're using this check, because in a quick test that I did I see that `ld` fails if there are unknown symbols, but your exit code check will catch this first:

```bash
    output=$(ld $file 2>&1) || exit_code=$?
    if [ $exit_code -ne 0 ]; then
        echo "Failed to execute ld: $output"
        exit 1
    fi

    if echo "$output" | grep "undefined reference"; then
        exit 1
    else
        echo "Verified ld status for $file"
    fi
```

```bash
alastair@pc:~$ gcc -fPIC -shared somefile.c -o somefile.so -nodefaultlibs -nostdlib
alastair@pc:~$ ld somefile.so
ld: warning: cannot find entry symbol _start; not setting start address
ld: somefile.so: undefined reference to `gzopen'
alastair@pc:~$ echo $?
1
```

In this case I wonder if it'd be better to use `nm` instead:
```bash
alastair@pc:~$ nm -D somefile.so
                 U gzopen
0000000000001030 T myfunction
```

In this case, `U` means "The symbol is undefined"

On a Mac, it looks like the equivalent is `nm -m`:
```
alastair@macos-12 ~ % nm -m somefile.dylib
                 (undefined) external _gzopen (dynamically looked up)
0000000000003f80 (__TEXT,__text) external _myfunction
```

Note that this even works on an x86 mac with an arm-compiled binary:
```
alastair@macos-12 ~ % file somefile.dylib
somefile.dylib: Mach-O 64-bit dynamically linked shared library arm64
```

Hope this is useful, let me know if I can help with adding these tests.

**Comment by dlemstra** [comment]
I think we should do this in a follow up PR but feel free to investigate this already. Thanks for all the help.

File: build/macos-arm64/build.libraries.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+$1/build.glib.sh
+$1/build.lqr.sh
+$1/build.pixman.sh
+$1/build.cairo.sh
+$1/build.fribidi.sh
+$1/build.harfbuzz.sh
+$1/build.pango.sh
+$1/build.croco.sh
+$1/build.rsvg.sh
+$1/build.raqm.sh
```
Comment (dlemstra) [inline build/macos-arm64/build.libraries.sh:34]: Not sure if GitHub is hiding this but does file follow the `insert_final_newline` rule in the `.editorconfig`?

File: build/macos-arm64/cross-compilation.meson [hunk 2]
```diff
@@ -0,0 +1,11 @@
+[binaries]
+pkgconfig = ['pkg-config']
+c = 'clang'
+cpp = 'clang++'
+objc = 'clang'
```
Comment (dlemstra) [inline build/macos-arm64/cross-compilation.meson:5]: Should this not be `clang++`?

File: build/macos-x64/install.dependencies.sh [hunk 3]
```diff
@@ -50,16 +50,16 @@ brew uninstall --force little-cms2
 brew uninstall --force libtiff
 brew uninstall --force jasper
 brew uninstall --force libjpeg
-brew uninstall --force jpeg
+brew uninstall --force jpeg-turbo
```
Comment (dlemstra) [inline build/macos-x64/install.dependencies.sh:53]: Should we not uninstall both of them?

File: build/macos-x64/install.dependencies.sh [hunk 4]
```diff
@@ -50,16 +50,16 @@ brew uninstall --force little-cms2
...
 brew uninstall --force sox
 brew uninstall --force libpng
 brew uninstall --force curl-openssl
 brew uninstall --force libzip
 brew uninstall --force httpd
-brew uninstall --ignore-dependencies brotli
-brew uninstall --ignore-dependencies libidn2
-brew uninstall --ignore-dependencies xz
-brew uninstall --ignore-dependencies zstd
+brew uninstall --force --ignore-dependencies brotli
```
Comment (dlemstra) [inline build/macos-x64/install.dependencies.sh:59]: This used to work without the `--force`? Was this added to be consistent with the rest? Do we really need it?

File: build/macos-x64/install.dependencies.sh [hunk 5]
```diff
@@ -5,7 +5,7 @@ export HOMEBREW_NO_AUTO_UPDATE=1
 
 brew list --formula
 
-brew install autoconf automake libtool nasm ossp-uuid pkg-config ragel
+brew install autoconf automake cmake libtool nasm ossp-uuid pkg-config ragel
```
Comment (dlemstra) [inline build/macos-x64/install.dependencies.sh:8]: Why was this added? This already worked?

**Review by dlemstra — COMMENTED** [review]
Thanks for all your help. Looking forward to getting this merged.

File: build/macos-arm64/build.libraries.sh (hunk 1 above)
Comment (alastairporter) [reply build/macos-arm64/build.libraries.sh:34]: there is a final newline here, I see it in my editor and on github

File: build/macos-arm64/cross-compilation.meson (hunk 2 above)
Comment (alastairporter) [reply build/macos-arm64/cross-compilation.meson:5]: I'm unsure here - I only filled it in because meson was complaining that it needed an `objc` entry here in order to do the cross compile, even though none of the libraries have objc code. I understand that this is all the same binary anyway, so maybe it's OK but if you can find a doc that says that clang++ is recommended for this then I can change it

File: build/macos-x64/install.dependencies.sh (hunk 5 above)
Comment (alastairporter) [reply build/macos-x64/install.dependencies.sh:8]: As I mentioned in the initial comment, I tested this first on a new macos VM and installed homebrew from scratch. In a default install, cmake isn't installed (it's installed by default on the github runners). I'm happy to remove this if we want the scripts to only run on the github runners, but in my experience if cmake is already installed this operation has no effect anyway

File: build/macos-x64/install.dependencies.sh (hunk 3 above)
Comment (alastairporter) [reply build/macos-x64/install.dependencies.sh:53]: I inspected the `brew list --formula` output at the beginning of this file and saw that the `jpeg` package wasn't actually installed, which is why I changed this instead of adding a new line. Happy to put both of them in for future-proofing if you want

File: build/macos-x64/install.dependencies.sh (hunk 4 above)
Comment (alastairporter) [reply build/macos-x64/install.dependencies.sh:59]: Same as the `cmake` change here - on my new VM that I set up these 4 packages don't exist after a fresh install of homebrew and so the `brew uninstall` command fails. `--force` ensures that the command runs even if the dependency isn't installed

File: build/macos-x64/install.dependencies.sh (hunk 3 above)
Comment (dlemstra) [reply build/macos-x64/install.dependencies.sh:53]: Aah okay, they probably changed this. We can keep it as is.

File: build/macos-arm64/cross-compilation.meson (hunk 2 above)
Comment (dlemstra) [reply build/macos-arm64/cross-compilation.meson:5]: I had to change this for the @ImageMagick project earlier but we can keep it like this if it works.

File: build/macos-x64/install.dependencies.sh (hunk 5 above)
Comment (dlemstra) [reply build/macos-x64/install.dependencies.sh:8]: I would rather only install things that we need to install on the runner. So please remove this.

File: build/macos-x64/install.dependencies.sh (hunk 4 above)
Comment (dlemstra) [reply build/macos-x64/install.dependencies.sh:59]: I would prefer to keep this as as. Since we are using the hosted GitHub actions builder.

**Comment by alastairporter** [comment]
I've made these changes + merged `main` into this branch to resolve the conflicts in the build scripts

**Review by dlemstra — APPROVED** [review]

**Comment by dlemstra** [comment]
Thanks for your help. And I just added the `hacktoberfest` to this repository in case you are part of this. And I hope you can help me with adding a verify stop for the MacOS builds.

**Comment by dlemstra** [comment]
And you might have noticed but I created a shared script to build the libraries. I waited with using this for the MacOS builds because of this PR. Feel free to make changes for that if you want to.

**Comment by alastairporter** [comment]
Great, thanks for merging this! This work was funded by [filestar](https://filestar.com/), but thanks for the offer to add the hacktoberfest tag as well!
I saw your changes to use the shared build script. I'll implement this in a new PR for the macos builds, and also take a look at the .dylib validation as we discussed.

Do you need any help with the magick.net side of things? I'm not a .net developer, but could try and give it a go if it would take another task off your hands.
I see that your verify step in magick.net actually runs `dotnet`, so this requires a machine of the same architecture as the library (and why you use qemu), is that right? Unfortunately without access to an M1 runner I don't think this will be possible to test.

**Comment by dlemstra** [comment]
Thanks for the offer to help @alastairporter and looking forward to your pull requests. I can do the Magick.NET side my side it requires quite a lot of changes in different spots. And yes we cannot test this yet because there are no hosted m1 runners for GitHub. For the Windows arm64 build I only build the code but I don't run the unit tests.
